subsys/dfu/dfu_target/src/dfu_target_smp.c

```diff
@@ -0,0 +1,345 @@
+int dfu_target_smp_write(const void *const buf, size_t len)
+{
+	int err = 0;
+	struct mcumgr_image_upload upload_res;
+
+	err = dfu_target_smp_enable();
+	if (err) {
+		return err;
+	}
+
+	err = img_mgmt_client_upload(&img_gr_client, buf, len, &upload_res);
+	if (err) {
+		return err;
+	}
+
+	return img_upload_res_update(&upload_res);
+}
+
+int dfu_target_smp_done(bool successful)
+{
+	int rc;
+	struct mcumgr_image_state res_buf;
+
+	rc = dfu_target_smp_enable();
+	if (rc) {
+		return rc;
+	}
+
+	if (successful) {
+		/* Read Image list */
+		LOG_INF("MCUBoot SMP image download ready.");
+		image_count = 0;
+		rc = img_mgmt_client_state_read(&img_gr_client, &res_buf);
+		if (rc) {
+			goto end;
+		}
+		rc = img_state_res_update(&res_buf);
+	} else {
+		LOG_INF("MCUBoot SMP image upgrade aborted %d", upload_state.image_num);
+	}
+end:
+	upload_state.image_size = 0;
+	upload_state.offset = 0;
+
+	return rc;
+}
+
+static struct mcumgr_image_data *discover_secondary_image_info(void)
+{
+	if (!image_count) {
+		return NULL;
+	}
+
+	for (int i = 0; i < image_count; i++) {
+		if (image_info[i].slot_num == 0) {
+			continue;
+		}
+		return &image_info[i];
+	}
+
+	return NULL;
+}
+
+int dfu_target_smp_schedule_update(int img_num)
+{
+	int err = 0;
+	struct mcumgr_image_state res_buf;
+	struct mcumgr_image_data *secondary_image;
+
+	/* Discover always Secondary image */
+	secondary_image = discover_secondary_image_info();
+	/* Test functionality here */
+	if (!secondary_image) {
+		LOG_ERR("Secondary Image info not available");
+		return -ENOENT;
+	}
+
+	err = dfu_target_smp_enable();
+	if (err) {
+		return err;
+	}
+
+	LOG_INF("MCUBoot image-%d upgrade scheduled. Reset device to apply",
+		secondary_image->slot_num);
+
+	err = img_mgmt_client_state_write(&img_gr_client, secondary_image->hash, false, &res_buf);
+	if (err) {
+		image_count = 0;
+		return err;
+	}
+
+	return img_state_res_update(&res_buf);
+}
+
+int dfu_target_smp_reset(void)
+{
+	int rc;
+
+	upload_state.image_size = 0;
+	upload_state.offset = 0;
+
+	rc = dfu_target_smp_enable();
+	if (rc) {
+		return rc;
+	}
+
+	if (dfu_target_reset_cb) {
+		/* MCUBoot not support Erase yet */
+		rc = 0;
+	} else {
+		rc = img_mgmt_client_erase(&img_gr_client, upload_state.image_num);
+	}
+
+	return rc;
+}
+
+int dfu_target_smp_reset_command(void)
+{
+	int rc;
+
+	upload_state.image_size = 0;
+	upload_state.offset = 0;
+
+	rc = dfu_target_smp_enable();
+	if (rc) {
+		return rc;
+	}
+
+	rc = os_mgmt_client_reset(&os_gr_client);
+	LOG_DBG("OS Reset command status:%d", rc);
+	if (rc) {
+		return rc;
+	}
+	recovery_mode_active = false;
+
+	return 0;
+}
+
+int dfu_target_smp_confirm_image(void)
+{
+	int rc;
+	int count = 0;
+	struct mcumgr_image_state res_buf;
+
+	if (dfu_target_reset_cb) {
+		/* Recovery mode can't confirm image  Application need to handle that */
+		return -EACCES;
+	}
+
+	while (1) {
+		if (count++ > 10) {
+			LOG_ERR("Image not swapped");
+			return -ENOENT;
+		}
+		k_sleep(K_SECONDS(1));
```

Comment:
Not sure on flow of this, swapping an image can take over a minute in some cases, if this function is used way after it has swapped then that is OK but if this is called right after reset, this is probably going to time out

Reply:
I simplify function just do the confirmation. User can poll state itself by read command.